File: harness/trace.py
"""Run the graph and keep everything it thought, not just what it concluded.

    python -m harness.trace                 all eight boards
    python -m harness.trace ground-stranded one of them

Writes `results/trace.json`: for every board, every node in the order it ran,
with the job it was given, the model's own reasoning, what it proposed, and —
at each turn of the loop — the gate's decision and the measurement behind it.

`tools/build_trace.py` turns that into a page. The point of keeping the
reasoning is that a review you cannot read the argument for is a review you have
to take on trust, which is the opposite of what this project is for.
"""

from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from console import utf8  # noqa: E402
from graph.build import MAX_PASSES, build_graph  # noqa: E402
from graph.prompts import BOARD_CONTEXT, SCHEMA, prompt_hash  # noqa: E402
from harness.checks import run_checks  # noqa: E402
from harness.distill import approx_tokens, distill  # noqa: E402
from harness.grade import grade, refuted  # noqa: E402
from harness.llm import PRICE_IN, PRICE_OUT, Client  # noqa: E402
from harness.ops import apply_edits, board_hash  # noqa: E402
from harness.presets import PRESETS, edits_for  # noqa: E402

OUT = ROOT / "results" / "trace.json"

#: What each node is for, in one line, for the page.
ROLES = {
    "ingest": "Parse, distil, and run the rule checks. No model is asked anything.",
    "datasheet": "What each pin is for, against what it is wired to.",
    "connections": "The wiring between parts, and off the board.",
    "layout": "Placement and copper.",
    "adjudicate": "Merge the three reviews, then let the board refute what it can.",
    "gate": "Stop, or go round again. Decided on measurements, never on the model's say-so.",
}
MEASURED = {"ingest", "gate"}


def job_of(prompt: str, distilled: str) -> str:
    """The prompt minus the board and the boilerplate every node shares.

    All six prompts carry the same 2,918-token board and the same schema. What
    differs between them is a paragraph, and that paragraph is the interesting
    part, so the page shows it and the board once.
    """
    job = prompt.replace(distilled, "").replace(BOARD_CONTEXT, "").replace(SCHEMA, "")
    job = job.replace("Here is the board.", "")
    job = job.replace(
        "Use the exact ref and net strings from the board so findings can be "
        "matched to\nthe design. Reply with JSON only, in exactly this shape:",
        "",
    )
    return "\n".join(line for line in job.splitlines() if line.strip()).strip()


def trace_board(name: str, title: str, board: dict, client: Client) -> dict:
    started = time.monotonic()
    state = build_graph(client).invoke({"board": board}, {"recursion_limit": 50})
    elapsed = time.monotonic() - started

    distilled = state["distilled"]
    rules = state.get("deterministic", [])

    steps = [
        {
            "node": "ingest",
            "pass": 0,
            "measured": True,
            "seconds": 0.0,
            "summary": (
                f"Distilled the board to {approx_tokens(distilled)} tokens. "
                + (
                    f"{len(rules)} rule check fired: "
                    + ", ".join(item["rule"] for item in rules)
                    if rules
                    else "No rule check fired."
                )
            ),
            "rules": rules,
        }
    ]

    gates = list(state.get("gates", []))
    for call in state.get("calls", []):
        node = call["node"]
        pass_no = int(call.get("label", "/pass1").split("pass")[-1] or 1)
        steps.append(
            {
                "node": node,
                "pass": pass_no,
                "measured": False,
                "job": job_of(call.get("prompt", ""), distilled),
                "reasoning": call.get("reasoning", ""),
                "answer": call.get("text", ""),
                "found": call.get("found", 0),
                "tokens_in": call.get("tokens_in", 0),
                "tokens_out": call.get("tokens_out", 0),
                "seconds": call.get("seconds", 0.0),
                "cached": bool(call.get("cached")),
            }
        )
        if node == "adjudicate":
            steps[-1]["dropped"] = [
                {"title": item["title"], "because": item["dropped"]}
                for item in state.get("dropped", [])
            ]
            steps[-1]["confirmed"] = [item["title"] for item in state.get("confirmed", [])]
            gate = next((g for g in gates if g["pass"] == pass_no), None)
            if gate:
                steps.append({**gate, "node": "gate", "measured": True})

    tokens_in = sum(s.get("tokens_in", 0) for s in steps)
    tokens_out = sum(s.get("tokens_out", 0) for s in steps)
    return {
        "board": name,
        "title": title,
        "board_hash": board_hash(board),
        "distilled": distilled,
        "distilled_tokens": approx_tokens(distilled),
        "steps": steps,
        "confirmed": state.get("confirmed", []),
        "proposed": len(state.get("findings", [])),
        "dropped": [
            {"title": d["title"], "because": d["dropped"]} for d in state.get("dropped", [])
        ],
        "refuted_reported": refuted(state.get("confirmed", []), board),
        "passes": state.get("passes", 1),
        "stopped": state.get("stopped", ""),
        "seconds": round(elapsed, 1),
        "tokens_in": tokens_in,
        "tokens_out": tokens_out,
        "dollars": round((tokens_in * PRICE_IN + tokens_out * PRICE_OUT) / 1e6, 5),
    }


def corpus(only: str = "") -> list[tuple[str, str, dict]]:
    good = json.loads((ROOT / "boards" / "stm32-good.json").read_text(encoding="utf-8"))
    out = [("clean", "The board as manufactured", good)]
    for preset in PRESETS:
        work = json.loads(json.dumps(good))
        apply_edits(work, edits_for(preset, work))
        out.append((preset["id"], preset["title"], work))
    if only:
        out = [row for row in out if row[0].startswith(only)]
        if not out:
            raise SystemExit(f"no board matching {only!r}")
    return out


def main() -> int:
    utf8()
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("board", nargs="?", default="")
    ap.add_argument("--tpm", type=int, default=40000)
    ap.add_argument("--concurrency", type=int, default=2)
    ap.add_argument("--out", type=Path, default=OUT)
    args = ap.parse_args()

    client = Client(concurrency=args.concurrency, tpm=args.tpm)
    boards = []
    for name, title, board in corpus(args.board):
        row = trace_board(name, title, board, client)
        boards.append(row)
        print(
            f"{name:<22} {row['passes']} pass · {len(row['steps'])} steps · "
            f"{row['stopped']:<26} {row['seconds']}s · ${row['dollars']}"
        )

    defects = {p["id"]: p for p in PRESETS}
    for row in boards:
        preset = defects.get(row["board"])
        row["defect"] = (
            {
                "id": preset["id"],
                "title": preset["title"],
                "breaks": preset["breaks"],
                "refs": preset["refs"],
                "nets": preset["nets"],
            }
            if preset
            else None
        )
        row["grade"] = grade(row["confirmed"], [row["defect"]] if preset else [])

    result = {
        "at": time.strftime("%Y-%m-%dT%H:%M:%S"),
        "model": client.model,
        "prompt_hash": prompt_hash(),
        "max_passes": MAX_PASSES,
        "roles": ROLES,
        "measured": sorted(MEASURED),
        "boards": boards,
        "totals": {
            "boards": len(boards),
            "calls": sum(len([s for s in b["steps"] if not s["measured"]]) for b in boards),
            "tokens_in": sum(b["tokens_in"] for b in boards),
            "tokens_out": sum(b["tokens_out"] for b in boards),
            "dollars": round(sum(b["dollars"] for b in boards), 4),
            "seconds": round(sum(b["seconds"] for b in boards), 1),
        },
    }
    args.out.parent.mkdir(parents=True, exist_ok=True)
    args.out.write_text(json.dumps(result, indent=1), encoding="utf-8")
    print(f"\n{result['totals']}")
    print(f"wrote {args.out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: site/trace.js
// The trace page: what the review graph did, and what it was thinking.
//
// One board at a time. The pipeline diagram shows the path that board actually
// took — including the loop back, when the gate sent it round again — and the
// steps below it carry the job each node was given, the model's own reasoning,
// and what it proposed.
//
// The two kinds of step are drawn differently on purpose. `ingest` and `gate`
// are measurements: they run rule checks and copper geometry and no model is
// asked anything. The three reviewers and the adjudicator are a model talking.
// Which is which is the whole argument, so the page says it in the design
// rather than only in the prose.

const TRACE = JSON.parse(document.getElementById("trace-data").textContent);
const $ = (id) => document.getElementById(id);

let current = TRACE.boards[0].board;

const esc = (v) =>
  String(v ?? "").replace(
    /[&<>"']/g,
    (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[c]
  );

const boardOf = (name) => TRACE.boards.find((b) => b.board === name);

// --------------------------------------------------------------- the diagram

const NODES = ["ingest", "datasheet", "connections", "layout", "adjudicate", "gate"];
const NODE_H = 34;
const GAP = 22;

function pipelineSvg(board) {
  const ran = new Set(board.steps.map((s) => s.node));
  const looped = board.passes > 1;
  const rows = NODES.map((node, i) => {
    const y = 8 + i * (NODE_H + GAP);
    const measured = TRACE.measured.includes(node);
    const count = board.steps.filter((s) => s.node === node).length;
    return { node, y, measured, count, on: ran.has(node) };
  });

  const height = 8 + NODES.length * (NODE_H + GAP);
  const boxes = rows
    .map(
      (r) => `
      <g class="pnode ${r.measured ? "measured" : "model"}${r.on ? " on" : ""}"
         data-node="${r.node}">
        <rect x="64" y="${r.y}" width="196" height="${NODE_H}" rx="3"></rect>
        <text x="76" y="${r.y + 21}">${r.node}</text>
        ${r.count > 1 ? `<text class="times" x="248" y="${r.y + 21}">×${r.count}</text>` : ""}
      </g>`
    )
    .join("");

  const arrows = rows
    .slice(0, -1)
    .map((r) => {
      const y = r.y + NODE_H;
      return `<path class="pedge" d="M 162 ${y} L 162 ${y + GAP - 4}"></path>
              <path class="pedge head" d="M 158 ${y + GAP - 8} L 162 ${y + GAP - 3} L 166 ${y + GAP - 8}"></path>`;
    })
    .join("");

  const gateY = rows[rows.length - 1].y + NODE_H / 2;
  const backY = rows[1].y + NODE_H / 2;
  const loop = `
    <path class="ploop${looped ? " on" : ""}"
          d="M 64 ${gateY} C 22 ${gateY}, 22 ${backY}, 60 ${backY}"></path>
    <path class="ploop head${looped ? " on" : ""}" d="M 55 ${backY - 4} L 62 ${backY} L 55 ${backY + 4}"></path>
    <text class="ploop-label${looped ? " on" : ""}" x="18" y="${(gateY + backY) / 2}"
          transform="rotate(-90 18 ${(gateY + backY) / 2})">
      ${looped ? "went round again" : "or round again"}
    </text>`;

  return `<svg viewBox="0 0 300 ${height}" class="pipeline"
               role="img" aria-label="The review graph, top to bottom">
    ${loop}${arrows}${boxes}
  </svg>`;
}

// ------------------------------------------------------------------- the run

function chip(board) {
  const defect = board.defect;
  const caught = board.grade.caught.length > 0;
  const state = !defect ? "clean" : caught ? "caught" : "missed";
  return `<button class="bchip ${state}${board.board === current ? " on" : ""}"
    data-board="${esc(board.board)}">
    <span class="bchip-name">${esc(board.board)}</span>
    <span class="bchip-meta">${board.passes} pass${board.passes === 1 ? "" : "es"} ·
      ${board.confirmed.length} reported</span>
  </button>`;
}

function stepBody(step) {
  if (step.node === "ingest") {
    return `<p class="step-summary">${esc(step.summary)}</p>
      ${step.rules.length
        ? `<ul class="rules">${step.rules
            .map(
              (r) => `<li><code>${esc(r.rule)}</code> ${esc(r.title)}
                <span class="why">${esc(r.why)}</span></li>`
            )
            .join("")}</ul>`
        : `<p class="muted">Nothing for the loop to chase, so the gate stops after
             one pass whatever the reviewers say.</p>`}`;
  }

  if (step.node === "gate") {
    const again = step.decision === "again";
    return `<p class="gate-decision ${again ? "again" : ""}">
        <code>${esc(step.decision)}</code>
        ${again ? '<span class="gate-tag">back to datasheet</span>' : ""}
      </p>
      <p class="step-summary">${esc(step.why)}</p>
      <dl class="facts">
        <dt>rules fired</dt><dd>${step.rules.length ? esc(step.rules.join(", ")) : "none"}</dd>
        <dt>still unaccounted</dt><dd>${step.unaccounted.length ? esc(step.unaccounted.join(", ")) : "none"}</dd>
        <dt>proposed / reported</dt><dd>${step.proposed} / ${step.confirmed}</dd>
      </dl>`;
  }

  const dropped = (step.dropped || []).length
    ? `<h4 class="eyebrow">Thrown out by the board, with no model consulted</h4>
       <ul class="dropped">${step.dropped
         .map(
           (d) => `<li><b>${esc(d.title)}</b><span class="why">${esc(d.because)}</span></li>`
         )
         .join("")}</ul>`
    : "";
  const kept = (step.confirmed || []).length
    ? `<h4 class="eyebrow">Reported</h4>
       <ul class="kept">${step.confirmed.map((t) => `<li>${esc(t)}</li>`).join("")}</ul>`
    : "";

  return `
    <details class="job">
      <summary>The job it was given</summary>
      <pre>${esc(step.job)}</pre>
    </details>
    <h4 class="eyebrow">Thinking</h4>
    <pre class="reasoning">${esc(step.reasoning) || "<span class='muted'>none returned</span>"}</pre>
    ${dropped}${kept}
    ${step.found !== undefined && !dropped && !kept
      ? `<p class="step-summary">Proposed ${step.found} finding${step.found === 1 ? "" : "s"}.</p>`
      : ""}`;
}

function renderBoard(name) {
  current = name;
  const board = boardOf(name);
  $("chips").innerHTML = TRACE.boards.map(chip).join("");
  for (const button of $("chips").querySelectorAll(".bchip")) {
    button.addEventListener("click", () => renderBoard(button.dataset.board));
  }

  $("pipeline").innerHTML = pipelineSvg(board);

  const defect = board.defect;
  $("board-head").innerHTML = `
    <h2>${esc(board.title)}</h2>
    ${defect
      ? `<p class="lede">${esc(defect.breaks)}</p>`
      : `<p class="lede">Nothing was changed — this is the false-alarm rate.</p>`}
    <dl class="facts">
      <dt>board hash</dt><dd>${esc(board.board_hash)}</dd>
      <dt>sent</dt><dd>${board.distilled_tokens} tokens</dd>
      <dt>passes</dt><dd>${board.passes} of ${TRACE.max_passes}</dd>
      <dt>stopped</dt><dd><code>${esc(board.stopped)}</code></dd>
      <dt>proposed → reported</dt><dd>${board.proposed} → ${board.confirmed.length}</dd>
      <dt>refuted by the board</dt><dd>${board.dropped.length} dropped,
        ${board.refuted_reported.length} still reported</dd>
      <dt>cost</dt><dd>$${board.dollars.toFixed(4)} · ${board.seconds}s ·
        ${board.tokens_in.toLocaleString()} in, ${board.tokens_out.toLocaleString()} out</dd>
    </dl>`;

  $("steps").innerHTML = board.steps
    .map((step, i) => {
      const measured = step.measured;
      const meta = measured
        ? "measured"
        : `${step.tokens_out.toLocaleString()} tokens out · ${step.seconds}s`;
      const looping = step.node === "gate" && step.decision === "again";
      return `<article class="step ${measured ? "measured" : "model"}${looping ? " looping" : ""}">
        <header>
          <span class="step-n">${String(i + 1).padStart(2, "0")}</span>
          <span class="step-node">${esc(step.node)}</span>
          ${step.pass ? `<span class="step-pass">pass ${step.pass}</span>` : ""}
          <span class="step-meta">${esc(meta)}</span>
        </header>
        <p class="role">${esc(TRACE.roles[step.node] || "")}</p>
        ${stepBody(step)}
      </article>`;
    })
    .join("");

  $("distilled").textContent = board.distilled;
  document.querySelector(".run").scrollTop = 0;
}

$("totals").textContent =
  `${TRACE.totals.boards} boards · ${TRACE.totals.calls} model calls · ` +
  `${(TRACE.totals.tokens_out / 1000).toFixed(0)}k tokens of output · ` +
  `$${TRACE.totals.dollars} · ${TRACE.totals.seconds}s`;
$("model-line").textContent = `${TRACE.model} · prompts ${TRACE.prompt_hash} · ${TRACE.at}`;

renderBoard(current);
